### ISSUE TITLE
TLS Error on any request

### ISSUE THREAD

**Issue by bernhardgessler** [gregsdennis/Manatee.Trello#200]
Seems to be a new thing. Any request seems to fail since a few days. Error:

> Could not create SSL/TLS secure channel

I had a similar issue before using another API of another service and it semed like thy had disconinued support for TLS 1.0
So I tried the same fix by adding
`ServicePointManager.SecurityProtocol = SecurityProtocolType.Ssl3 | SecurityProtocolType.Tls12;`
in the constructor of Manatee.Trello.Rest.WebApiClient
This seems to resolve this problem. But it only compiles for .net 4.5 and not for .Net Standard. I don't have much experience with the latter one so I'm not filing a pull request this time.

**Comment by bernhardgessler** [comment]
Did some more research and noticed the .Net 4.6+ already supports TLS 1.2 by default. Same for .Net Starndard 1 & 2. So my issue only comes up because I'm compiling on .Net 4.5.

I'm currently going with the workaround to add this code to my app initialization:
`// TODO: Remove after upgtrade to .Net 4.6+
ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls12;`

Manatee.Trello may not need any adjustments.

**Comment by bernhardgessler** [comment]
Just checked but it seems like your test projects are using 4.6.2, while Manatee.Trello uses 4.5. So when running the test you may not encounter any issues.

I downgraded your test projects to 4.5 and the same issue occurs.

Using 4.5 myself I'm highly in favor of keeping compatibility with 4.5 for now. Maybe we could add this to the contructor of the WebApiClient for now:
`#if NET45`
`ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls12;`
`#endif`

Some more info I found on TLS support in .Net:
https://blogs.perficient.com/2016/04/28/tsl-1-2-and-net-support/

On a related note: A colleage just approached me, saying his Trello app on Android stopped synchronizing. He's using Android 4.2 which according to quick resaearch doesn't support TLS 1.2. So I guess its confirmed that they rolled out a bigger change.

**Comment by gregsdennis** [comment]
Okay... Thank for the heads up.  I have another update I'm working on.  I'll include this in with that.

**Comment by gregsdennis** [comment]
Thinking about this again, I'm having difficulty justifying putting this in the library.

The client needs it to work for .Net Framework 4.5, so from that point of view, sure.

On the other hand, it's a global setting.  It's likely that a process making a call to Trello is also making other calls.  Setting this in the Trello library seems presumptuous about the client.  I'm not sure its the responsibility of this library to update a system setting like this.

I think that this should be something that's set in an application startup rather than something set by a library.  (It doesn't have to be set in the constructor, and it only has to be set once.)

This would shift the responsibility of setting this to the client (you), which I think is more proper.

**Comment by gregsdennis** [comment]
I've added a comment on the *Getting Started* and *Configuration* pages.  You can see it at the top [here](https://github.com/gregsdennis/Manatee.Trello/blob/bug/177-collection-limit/Docs/2.%20Configuration.md)

**Comment by bernhardgessler** [comment]
Double checked the docs. It's not a change by Microsoft. It's been a change on the Trello API about 24 hours ago.. 
I see your thoughts and they are very reasonable. What I like about this line
`ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls12;`
though compared to my initial approach
`ServicePointManager.SecurityProtocol = SecurityProtocolType.Ssl3 | SecurityProtocolType.Tls12;`
Is that it doesn't override any settings, just adds TLS 1.2. as a possible security standard to be used for making connections. I don't see how this would break anything. But you're right, there might be cases I'm not able to foresee.

Whatever you choose to do, I'm fine with it. Let's just rephrase the docs a bit. I suggest:

> Trello recently discontinued supporting TLS 1.0 which up to .Net 4.5 is the default for secure connections. If you are using .Net 4.6 or newer you are good already. If you use .Net 4.5, please put the following line of code in your application startup logic:
`ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls12;`

For more information regarding TLS support in .Net, please see this blog post.

**Comment by gregsdennis** [comment]
I see your point.  I didn't realize it was a Trello support issue.  In that case, I don't mind adding the code to the library.  I've also pushed this up on their slack community to get their input.

**Comment by gregsdennis** [comment]
Sooo....  They just posted this in response to my question: https://trello.com/c/8np6xdTH/16-removing-tls-10-support

**Comment by gregsdennis** [comment]
@bernhardgessler in the meantime, are you good with just setting that in your code?

**Comment by bernhardgessler** [comment]
Works for me. That's what I'm doing right now.
Thanks for your effort. I love your library!

**Comment by TimSirmovics** [comment]
We also had this issue the other day, i'm sure along with many others. We added the relevant `SeucrityProtocol` line into `Global.asax`.

**Comment by gregsdennis** [comment]
@bernhardgessler please watch for v3.0.7.  This should take care of that (though also having that line in your code won't hurt anything, I suppose).

**Comment by ganga0810** [comment]
I was having same issue in my Asp.Net and solved by adding `ServicePointManager.SecurityProtocol = SecurityProtocolType.Tls12;` to Global.asax. Now I am having the same issue in my Xamarin Forms app. Could you please guide me where should I put this line? I have tried to put in App.cs, MainActivity.cs but no luck. I also tried to upgrade to 3.0.7 and that error went away but now I am getting `System.TypeLoadException: Could not load type of field 'JobsheetAdvationXamarin.JobStatus+<btnsave_Clicked>d__11:<newcard>5__11' (15) due to: Could not resolve type with token 010000b1 (from typeref, class/assembly Manatee.Trello.ICard, Manatee.Trello, Version=3.0.0.0, Culture=neutral, PublicKeyToken=f502fcc17fc907d6) assembly:Manatee.Trello, Version=3.0.0.0, Culture=neutral, PublicKeyToken=f502fcc17fc907d6 type:Manatee.Trello.ICard member:<none>`. Please guide me. Thanks

**Comment by gregsdennis** [comment]
I've incorporated this into the lib as of v3.0.7. You shouldn't need to do anything. Which version of Manatee.Trello are you using?

**Comment by ganga0810** [comment]
I was using 2.4.0 which worked perfectly until few days ago. Now I have updated to ver 3.0.7 which I am getting the exception I posted above.

**Comment by ganga0810** [comment]
do you have any idea where I could state the SecurityProtocolType.Tls12 in Xamarin Forms application?

**Comment by gregsdennis** [comment]
I think the problem is Xamarin.  I don't specifically target Mono.  Can you tell me which target you're using (Framework 4.5 / Standard 1.3 / Standard 2.0)?

The fix was only applied for Fwk 4.5 because 4.6+ and Standard already has this enabled.

I would suggest that you put it wherever your configuration or bootstrapping code is.  Also I suggest you use the bit-wise OR operator so that you don't overwrite values already in that field.

```csharp
ServicePointManager.SecurityProtocol |= SecurityProtocolType.Tls12;
```
